mockrxandroidble/src/main/java/com/polidea/rxandroidble/mockrxandroidble/RxBleConnectionMock.java

```diff
@@ -0,0 +1,190 @@
+        return Observable.just(rssi);
+    }
+
+    @Override
+    public Observable<Observable<byte[]>> setupNotification(@NonNull UUID characteristicUuid) {
+        final Observable<Observable<byte[]>> availableObservable = notificationObservableMap.get(characteristicUuid);
+
+        if (availableObservable != null) {
+            return availableObservable;
+        }
+
+        final Observable<Observable<byte[]>> newObservable = createCharacteristicNotificationObservable(characteristicUuid)
+                .doOnUnsubscribe(() -> dismissCharacteristicNotification(characteristicUuid))
+                .map(notificationDescriptorData -> observeOnCharacteristicChangeCallbacks(characteristicUuid))
+                .replay(1)
+                .refCount();
+        notificationObservableMap.put(characteristicUuid, newObservable);
+
+        return newObservable;
+    }
+
+    @Override
+    public Observable<BluetoothGattCharacteristic> writeCharacteristic(@NonNull BluetoothGattCharacteristic bluetoothGattCharacteristic) {
+        return getCharacteristic(bluetoothGattCharacteristic.getUuid())
+                .map(characteristic -> characteristic.setValue(bluetoothGattCharacteristic.getValue()))
+                .flatMap(ignored -> Observable.just(bluetoothGattCharacteristic));
+    }
+
+    @Override
+    public Observable<byte[]> writeCharacteristic(@NonNull UUID characteristicUuid, @NonNull byte[] data) {
+        return getCharacteristic(characteristicUuid)
+                .map(characteristic -> characteristic.setValue(data))
+                .flatMap(ignored -> Observable.just(data));
+    }
+
+    @Override
+    public Observable<byte[]> writeDescriptor(UUID serviceUuid, UUID characteristicUuid, UUID descriptorUuid, byte[] data) {
+        return discoverServices()
+                .flatMap(rxBleDeviceServices -> rxBleDeviceServices.getDescriptor(serviceUuid, characteristicUuid, descriptorUuid))
+                .map(bluetoothGattDescriptor -> bluetoothGattDescriptor.setValue(data)).flatMap(ignored -> Observable.just(data));
+    }
+
+    private Observable<Observable<byte[]>> createCharacteristicNotificationObservable(UUID characteristicUuid) {
+        return getClientConfigurationDescriptor(characteristicUuid)
+                .flatMap(bluetoothGattDescriptor -> setupCharacteristicNotification(bluetoothGattDescriptor, true))
+                .flatMap(bluetoothGattDescriptorPair -> Observable.create(subscriber -> subscriber.onNext(bluetoothGattDescriptorPair)))
+                .flatMap(bluetoothGattDescriptorPair -> {
+                    if (!characteristicNotificationSources.containsKey(characteristicUuid)) {
+                        return Observable.error(new IllegalStateException("Lack of notification source for given characteristic"));
+                    }
+                    return Observable.just(characteristicNotificationSources.get(characteristicUuid));
+                })
+                .doOnUnsubscribe(() -> {
```

Comment:
Looks like whole doOnUnsubscribe block is a duplicate (look at dismissCharacteristicNotification and its use)